p2p/peering.go

```diff
@@ -1107,6 +1115,44 @@ func (rp *RemotePeer) receivedGetData(ctx context.Context, msg *wire.MsgGetData)
 	}
 }
 
+func (rp *RemotePeer) addRequestedInitState(c chan<- *wire.MsgInitState) (newRequest bool) {
+	rp.requestedInitStateMu.Lock()
+	if rp.requestedInitState != nil {
+		rp.requestedInitStateMu.Unlock()
+		return false
+	}
+	rp.requestedInitState = c
+	rp.requestedInitStateMu.Unlock()
+	return true
+}
+
+func (rp *RemotePeer) deleteRequestedInitState() {
+	rp.requestedInitStateMu.Lock()
+	rp.requestedInitState = nil
+	rp.requestedInitStateMu.Unlock()
+}
+
+func (rp *RemotePeer) receivedInitState(ctx context.Context, msg *wire.MsgInitState) {
+	const opf = "remotepeer(%v).receivedInitState"
+	rp.requestedInitStateMu.Lock()
+
+	if rp.requestedInitState == nil {
+		op := errors.Opf(opf, rp.raddr)
+		err := errors.E(op, errors.Protocol, "received unrequested init state")
+		rp.Disconnect(err)
+		rp.requestedInitStateMu.Unlock()
```

Comment:
Swap these two lines, so that you do the Disconnect() after releasing the mutex. 

There's no obvious race doing the current way, but in general you should try to do as little as possible inside a critical section (that is, while having a locked mutex), for safety and future proofing.

In fact, you can rewrite the use of the `requestedInitStateMu` mutex in this function to keep it really short:

```go
rp.requestedInitStateMu.Lock()
c := rp.requestedInitState
rp.requestedInitState = nil
rp.requestedInitStateMu.Unlock()
if c == nil { 
// error out
}
```

Writing unconditionally to the field (`rp.requestedInitState = nil`) is fine because if it's already nil, it will remain so and if it's non-nil, `c` will have the channel for the later code to act upon. This makes the time spent in the critical section the smallest possible and makes it easier to deduce there will be no deadlocks possible (because there are no function calls that could themselves trigger a deadlock).